README.md
#OnioNS - the Onion Name System
### Tor-Powered Distributed DNS for Tor Hidden Services

OnioNS is a distributed, privacy-enhanced, metadata-free, and highly usable DNS for Tor hidden services. OnioNS allows hidden service operators to select a meaningful and globally-unique domain name for their service, which users can then reference from the Tor Browser. The system is powered by the Tor network, relies on a distributed database, and provides anonymity to both operators and users. This project aims to address the major usability issue that has been with Tor hidden services since their introduction in 2002.

### Repository Details [![Build Status](https://travis-ci.org/Jesse-V/OnioNS-HS.svg?branch=json-rpc)](https://travis-ci.org/Jesse-V/OnioNS-HS)

This repository provides functionality for hidden service operator, such as the capability to claim an OnioNS domain name.

### Supported Systems

**Debian 7 and 8, Ubuntu 14.04 - 15.10, Mint 17 - 17.2, Fedora 21 - 23**

Please see the [OnioNS-common README](https://github.com/Jesse-V/OnioNS-common#supported-systems) for more information.

### Installation

There are several methods to install the OnioNS software. The method of choice depends on your system. If you are on Ubuntu or an Ubuntu-based system (Lubuntu, Kubuntu, Mint) please use the PPA method. If you are running Debian Wheezy, please use the .deb method. Otherwise, for all other distributions, please install from source.

* **Install from PPA**

> 1. **sudo add-apt-repository ppa:jvictors/tor-dev**
> 2. **sudo apt-get update**
> 3. **sudo apt-get install tor-onions-hs**

This is the recommended method as it's very easy to stay up-to-date with my releases.

* **Install from .deb file**

I provide builds for Debian Wheezy in the [Releases section](https://github.com/Jesse-V/OnioNS-HS/releases) for several architectures. For other architectures, you may download from [my PPA](https://launchpad.net/~jvictors/+archive/tor-dev/+packages).

* **Install from source**

> 1. Install tor-onions-common by following [these instructions](https://github.com/Jesse-V/OnioNS-common#installation).
> 2. Download and extract the latest release from the [Releases page](https://github.com/Jesse-V/OnioNS-HS/releases).
> 3. *(mkdir build; cd build; cmake ../src; make; sudo make install)*

If you are actively developing OnioNS, I have actively prepared two scripts, devBuild.sh and checkBuild.sh. Please see them for more information.

You can cleanup your build with **rm -rf build**

### Usage

The hidden service edition of the OnioNS software is primarily a command-line utility. Currently it supports the ability to claim a domain name, a number of sub-domains, and send this information over a Tor circuit to the OnioNS network.

> 1. Create a hidden service. See [this guide](https://www.torproject.org/docs/tor-hidden-service) for instructions on how to do this. As this is beta software, I do not recommend using a valuable hidden service in production, so I recommend making a new one for testing purposes.
> 2. **onions-hs --hsKey /var/lib/tor/hidden_service/private_key**
> 3. Follow the on-screen instructions. You may claim a single .tor domain (which will point to your hidden service) and up to 24 subdomains, which must point to a .tor or .onion address of your choosing.

A manpage is available for your convenience.

#### Minimum System Requirements

1 CPU core, 1 MB of free disk space, and N * 128 MB of available RAM where N is the number of available CPU cores. Please add "-w 1" if you have more than one CPU core but less than N * 128 MB available RAM. This can sometimes be the case if your HS is hosted on a minuscule VM.

### Bug Reporting

Please open a ticket on Github. If you do not have a Github account, please contact kernelcorn on #tor-dev on OFTC IRC, or email kernelcorn at riseup dot net. Please follow the same process for filing enhancement requests. I use PGP key 0xC20BEC80. I accept pull requests if you want to contribute.


debian/changelog
tor-onions-hs (0.7.1.2~trusty) trusty; urgency=medium

  * UX: Introduce main menu.
  * UX: Display master EdDSA key as a series of 16 words.
  * UX: Used colorized output to terminal.
  * Security: Use master EdDSA key to manage record.
  * Security: Stop loading/saving keys to disk.
  * Include american-english-small wordlist from wamerican-small/testing.
  * Sign Record using EdDSA and RSA keys instead of Record doing that.
  * Initial setup for proof-of-work.

 -- Jesse Victors <kernelcorn@torproject.org>  Sun, 02 Oct 2016 12:00:00 -0500

tor-onions-hs (0.6.2.1~trusty) trusty; urgency=medium

  * Fix RPATH assignment to allow the Linux loader to find OnioNS-common.

 -- Jesse Victors <kernelcorn@torproject.org>  Wed, 30 Sep 2015 19:00:00 -0800

tor-onions-hs (0.6.1.5~trusty) trusty; urgency=medium

  * Use authenticated connections to Quorum node.
  * Add Clang's scan-build static analysis tool to build scripts.
  * Install to /usr/local/ when compiled from source.
  * Re-fixed missing linkage against pthread, necessary for armhf builds.

 -- Jesse Victors <kernelcorn@torproject.org>  Mon, 28 Sep 2015 13:00:00 -0800

tor-onions-hs (0.6.0.3~trusty) trusty; urgency=medium

  * Save newly-generated Record to disk before upload. (#8)
  * Send Records to Quorum nodes hosted by hidden services.
  * Remove dependency on Tor. (#3)
  * Introduce flag to specify Tor's SOCKS port. (#5)
  * Introduce flag to specify number of worker threads. (#4)

 -- Jesse Victors <kernelcorn@torproject.org>  Wed, 26 Aug 2015 23:00:00 -0800

tor-onions-hs (0.5.0.0~trusty) trusty; urgency=medium

  * Reduced most compiler and static analysis warnings.
  * Added popt-dev as a missing dependency.
  * Updated network protocol.

 -- Jesse Victors <kernelcorn@torproject.org>  Fri, 07 Aug 2015 16:00:00 -0800

tor-onions-hs (0.4.0.2~trusty) trusty; urgency=medium

  * Automatic uploading of new Records to a Quorum server.
  * Clarified many registration prompts.
  * Added dependency for a system installation of Tor.
  * Standardized the build process.
  * Replaced SourceForge tclap library with standard Linux popt.
  * Created onions-hs manpage.
  * Standardized the build process.
  * Improved git commit styles to match common standard.
  * Clarified README.
  * Many bug fixes.

 -- Jesse Victors <kernelcorn@torproject.org>  Tues, 04 Aug 2015 09:00:00 -0800

tor-onions-hs (0.3.2.6~trusty) trusty; urgency=medium

  * Fixed missing dependency on tor-onions-common.
  * Fixed missing linkage against pthread, necessary for armhf builds.

 -- Jesse Victors <kernelcorn@torproject.org>  Mon, 06 Jul 2015 01:00:00 -0400

tor-onions-hs (0.3.2.4~trusty) trusty; urgency=medium

  * Initial build against tor-onions-common library.
  * Fixed compilation against tor-onions-common headers.

 -- Jesse Victors <kernelcorn@torproject.org>  Sun, 5 Jul 2015 15:31:00 -0400

tor-onions-hs (0.3.1.2~trusty) trusty; urgency=medium

  * Initial build of hidden service functionality.
  * Removed debugging symbols.
  * Fixed build dependencies.

 -- Jesse Victors <kernelcorn@torproject.org>  Thur, 01 Jul 2015 22:45:00 -0400


src/RecordManager.cpp

#include "RecordManager.hpp"
#include "Terminal.hpp"
//#include <onions-common/Config.hpp>
//#include <onions-common/Utils.hpp>
#include <botan/auto_rng.h>
#include <functional>
#include <iostream>
#include <iomanip>
#include <fstream>
#include <sstream>


const size_t WORD_32 = 2 << 16;
const int LINE_LENGTH = 80;

const Terminal::Style bold(Terminal::BOLD);
const Terminal::Style no_bold(Terminal::NO_BOLD);
const Terminal::Style cyan(Terminal::CYAN, Terminal::LIGHT);
const Terminal::Style red(Terminal::RED, Terminal::LIGHT);
// const Terminal::Style green(Terminal::GREEN, Terminal::LIGHT);
// const Terminal::Style yellow(Terminal::YELLOW, Terminal::LIGHT);
const Terminal::Style blue(Terminal::BLUE, Terminal::LIGHT);
const Terminal::Style reset(Terminal::DEFAULT);


#ifndef INSTALL_PREFIX
#error CMake has not defined INSTALL_PREFIX!
#endif


void RecordManager::mainMenu() const
{
  std::cout << std::endl << bold << cyan;
  std::cout << "Welcome to the OnioNS name management portal!" << std::endl;
  std::cout << reset << no_bold << blue;

  while (true)
  {
    switch (showMenu())
    {
      case 1:
        registerName();
        break;

      case 2:
        modifyName();
        break;

      case 3:
        printHelp();
        break;

      case 4:
        exit(EXIT_SUCCESS);
    }
  }
}



RecordPtr generateRecord()
{
  std::string type = "Create";
  std::string name = "example.tor";
  std::string pgp = "1234";
  StringMap subdomains;
  uint32_t rng = 0, nonce = 0;
  return std::make_shared<Record>(type, name, pgp, subdomains, rng, nonce);
}



void RecordManager::registerName() const
{
  EdDSA_KEY edSecKey = generateSecretKey();
  auto wordList = getWordList();
  auto list = toWords(wordList, edSecKey);

  const int PRINT_SIZE = 4;
  for (int j = 0; j < PRINT_SIZE; j++)
  {
    std::cout << "  ";
    for (int k = 0; k < PRINT_SIZE; k++)
      std::cout << std::left << std::setw(19) << list[j * PRINT_SIZE + k];
    std::cout << std::endl << std::endl;
  }



  RecordPtr record = generateRecord();
  std::cout << *record << std::endl;
  // record is missing edKey, serviceKey, serviceSignature, edSig

  /*
  std::string resolve(const std::string&) const;
    Botan::SecureVector<uint8_t> hash() const;
    uint32_t computePOW(const std::vector<uint8_t>&) const;
    bool computeValidity() const;
    std::string computeOnion() const;
    std::vector<uint8_t> asBytes(bool forSigning = false) const;
    Json::Value asJSON() const;
    friend std::ostream& operator<<(std::ostream&, const Record&);
    */



  /*
    Botan::AutoSeeded_RNG rng;
    Botan::PK_Signer signer(*privateKey_, "EMSA4(SHA-512)");
    auto bytes = r.getServiceSigningScope();
    auto sig = signer.sign_message(bytes, jsonStr.size(), rng);
    std::copy(sig.begin(), sig.end(), signature_.begin());
  */

  // ed25519_sign(message, message_len, sk, pk, signature);

  /*
  std::string bytes = asBytes(); // last 4 bytes is the nonce
  bytes.data()[size - 4]++;
  setSignature(sign(privateKey, bytes.data()))
  auto val = computePOW();
  if (val < best)
    bestNonce = nonce;
*/
}



void RecordManager::modifyName() const
{
  std::cout << "This functionality has not yet been implemented." << std::endl;
}



void RecordManager::printHelp() const
{
  // clang-format should hard-wrap and insert quotes into the below text
  std::vector<std::string> text;

  text.push_back(std::string(
      "The Onion Name System (OnioNS) is a privacy-enhanced, metadata-free, "
      "and highly-usable DNS for Tor onion services. Using this software, you "
      "can anonymously register a meaningful and globally-unique domain name "
      "for your service. OnioNS is built on top of the Tor network, does not "
      "require any modifications to the Tor binary, and there are no central "
      "authorities in charge of the domain names. This project was "
      "specifically engineered to solve the usability problem with onion "
      "services."));

  text.push_back(std::string(
      "We have designed the Onion Name System (OnioNS) with security and "
      "privacy in mind. Unlike the Internet DNS, OnioNS does not ask for any "
      "personal information when you register a name for your service. "
      "Instead, we use a combination of asymmetric cryptographic keys to map "
      "your onion/hidden service's RSA key to a meaningful name. Specifically, "
      "your RSA key digitally signs a master edDSA key. The master key in turn "
      "signs the RSA key, a meaningful name, subdomains, and other data. This "
      "way, everyone can verify that the name belongs to you."));

  text.push_back(std::string(
      "In effect, you use the master key to have complete control over your "
      "name record. Once you have registered a name on your own, you will be "
      "prompted to distribute this record to specific nodes inside the Tor "
      "network. If accepted by the network, users will be able to type the "
      "meaningful name into the Tor Browser and load your onion service. The "
      "digital signatures ensure authenticity and prevent anyone from "
      "manipulating your record without your permission."));

  text.push_back(std::string(
      "It is very important that you keep the master key safe. This portal "
      "will translate the master key into a series of words, which will make "
      "it easier for you to write down and retrieve. You can keep these words "
      "offline and air-gapped. This software will never write the master key "
      "to disk and will securely erase it from RAM when the program exits. "
      "We cannot recover lost keys, restore control of your record if the "
      "master key has been compromised, or hand your key over to any third "
      "party."));

  text.push_back(std::string(
      "Online documentation is coming soon. We assume that this software will "
      "be run offline or on a headless environment, so we will continue to "
      "provide the necessary documentation within this software package."));

  printParagraphs(text);
}



int RecordManager::showMenu() const
{
  std::cout << blue << std::endl;
  std::cout << "Main menu. Select from one of the options below:\n";
  std::cout << reset;
  std::cout << "1) Register a new name \n";
  std::cout << "2) Modify an existing name record \n";
  std::cout << "3) Get help and learn more \n";
  std::cout << "4) Close this application \n";
  std::cout << blue;

  return readInt("Your selection:", "That is not a valid selection.",
                 [](int i) { return i >= 0 && i <= 4; });
  ;
}



std::vector<std::string> RecordManager::toWords(
    const std::vector<std::string>& wordList,
    const EdDSA_KEY& sKey) const
{
  std::vector<std::string> results;
  if (sKey.size() != Const::EdDSA_KEY_LEN)
    return results;

  for (int j = 0; j < Const::EdDSA_KEY_LEN; j += 2)
  {
    uint16_t value = *reinterpret_cast<const uint16_t*>(sKey.data() + j);
    results.push_back(wordList[value]);
  }

  return results;
}



EdDSA_KEY RecordManager::generateSecretKey() const
{
  static Botan::AutoSeeded_RNG rng;
  EdDSA_KEY key;
  rng.randomize(key.data(), Const::EdDSA_KEY_LEN);
  return key;
}



std::vector<std::string> RecordManager::getWordList() const
{
  std::ifstream file;
  std::string path = INSTALL_PREFIX + "/share/onions-hs/american-english-small";
  file.open(path, std::fstream::in);

  std::vector<std::string> results;
  if (file.is_open())
  {
    std::string line;
    while (file.good())
    {
      getline(file, line);
      results.push_back(line);
    }
    results.pop_back();

    for (int j = -7000; results.size() < WORD_32; j++)
      results.push_back(std::to_string(j));

    file.close();
  }
  else
    std::cerr << "Failed to open " << path << std::endl;

  return results;
}



int RecordManager::readInt(const std::string& prompt,
                           const std::string& failMessage,
                           std::function<bool(int)> check) const
{
  int choice = 0;
  std::string line;
  while (true)
  {
    Terminal::clearLine();
    std::cout << blue << prompt << " " << reset;
    getline(std::cin, line);
    if (std::istringstream(line) >> choice)
    {
      if (check(choice))
        break;
      else
        std::cout << red << failMessage << std::endl;
    }
    else
      std::cout << red << "Please provide a numeric value." << std::endl;
  }
  std::cout << std::endl;

  return choice;
}



void RecordManager::printParagraphs(std::vector<std::string>& text) const
{
  // print each paragraph, wrap a LINE_LENGTH chars
  std::for_each(text.begin(), text.end(), [](std::string& paragraph) {
    int index = LINE_LENGTH;
    while (index < paragraph.size())
    {
      auto pos = paragraph.find_last_of(' ', index);
      paragraph.replace(pos, 1, 1, '\n');
      index = pos + LINE_LENGTH;
    }
    std::cout << paragraph << std::endl << std::endl;
  });
}


/*
for (int j = 999999; j > 0; j--)
{
  printLine("Going " + std::to_string(j));
  clearLine();
  // std::this_thread::sleep_for(std::chrono::milliseconds(250));
}

RecordPtr HS::promptForRecord()
{
std::cout
    << "Here you can claim a .tor domain and multiple subdomains for your"
       " hidden service. They can point to either a .tor or a .onion domain,"
       " keeping it all within Tor. For example, you may claim"
       " \"example.tor\" -> \"onions55e7yam27n.onion\", \"foo.example.tor\""
       " -> \"foo.tor\", \"bar.example.tor\" -> \"bar.tor\", and"
       " \"a.b.c.example.tor\" -> \"3g2upl4pq6kufc4m.onion\"."

std::string name;
std::cout << "\nThe primary domain name must end in \".tor\"" << std::endl;
while (name.length() < 5 || !Utils::strEndsWith(name, ".tor"))
{
  std::cout << "The primary domain name: ";
  std::getline(std::cin, name);
  name = Utils::trimString(name);
}

std::string pgp;
  std::cout << "Your PGP fingerprint: ";
  std::getline(std::cin, pgp);  //"AD97364FC20BEC80"

  pgp = Utils::trimString(pgp);
  // if (pgp.empty())
  //  break;
}

std::cout << "\nYou may provide up to 24 subdomain-destination pairs.\n"
             "These must end with \"."
          << name << "\". Leave the "
                     "subdomain blank when finished."
          << std::endl;

StringMap list;
for (int n = 0; n < 24; n++)
{
  std::string src = name, dest;

  while (!Utils::strEndsWith(src, "." + name))
  {
    std::cout << "Subdomain " << (n + 1) << ": ";
    std::getline(std::cin, src);
    src = Utils::trimString(src);

    if (src.length() == 0)
      break;
  }

  if (src.length() == 0)
    break;

  while ((!Utils::strEndsWith(dest, ".tor") &&
          !Utils::strEndsWith(dest, ".onion")) ||
         (Utils::strEndsWith(dest, ".onion") && dest.length() != 16 + 6))
  {
    std::cout << "   Destination: ";
    std::getline(std::cin, dest);
    dest = Utils::trimString(dest);
  }

  src.erase(src.find("." + name));
  list.push_back(std::make_pair(src, dest));
  std::cout << std::endl;
}

std::cout << std::endl;
// auto r = std::make_shared<CreateR>(Utils::loadKey(keyPath_), name, pgp);
// r->setSubdomains(list);

// return r;
return nullptr;
}
*/
